Source/Core/DolphinQt/Debugger/CodeDiffDialog.cpp

```diff
@@ -0,0 +1,406 @@
+  std::vector<Diff>().swap(m_exclude);
+  JitInterface::SetProfilingState(JitInterface::ProfilingState::Disabled);
+}
+
+void CodeDiffDialog::ClearBlockCache()
+{
+  Core::State old_state = Core::GetState();
+
+  if (old_state == Core::State::Running)
+    Core::SetState(Core::State::Paused);
+
+  JitInterface::ClearCache();
+
+  if (old_state == Core::State::Running)
+    Core::SetState(Core::State::Running);
+}
+
+void CodeDiffDialog::OnRecord(bool enabled)
+{
+  JitInterface::ProfilingState state;
+
+  if (enabled)
+  {
+    ClearBlockCache();
+    m_record_btn->setText(tr("Recording..."));
+    state = JitInterface::ProfilingState::Enabled;
+    m_exclude_btn->setEnabled(true);
+    m_include_btn->setEnabled(true);
+  }
+  else
+  {
+    ClearBlockCache();
+    m_record_btn->setText(tr("Start Recording"));
+    state = JitInterface::ProfilingState::Disabled;
+    m_exclude_btn->setEnabled(false);
+    m_include_btn->setEnabled(false);
+  }
+
+  m_record_btn->update();
+  JitInterface::SetProfilingState(state);
+}
+
+void CodeDiffDialog::OnInclude()
+{
+  const auto recorded_symbols = CalculateSymbolsFromProfile();
+  if (m_include.empty() && m_exclude.empty())
+  {
+    m_include = recorded_symbols;
+    return;
+  }
+
+  std::vector<Diff> current_diff;
+  if (m_exclude.empty())
+  {
+    current_diff = recorded_symbols;
+  }
+  else
+  {
+    for (auto& iter : recorded_symbols)
+    {
+      auto pos = lower_bound(m_exclude.begin(), m_exclude.end(), iter.symbol, AddrOP);
+
+      if (pos->symbol != iter.symbol)
+      {
+        current_diff.push_back(iter);
+      }
+    }
+  }
+
+  if (!m_include.empty())
+  {
+    m_include.erase(std::remove_if(m_include.begin(), m_include.end(),
+                                   [&](Diff& v) {
+                                     // Remove values if it can't find matching symbols
+                                     return std::none_of(
+                                         current_diff.begin(), current_diff.end(),
+                                         [&](Diff& p) { return p.symbol == v.symbol; });
+                                   }),
+                    m_include.end());
+  }
+  else
+  {
+    m_include = recorded_symbols;
+    RemoveMatchingSymbolsFromIncludes(m_exclude);
+  }
+}
+
+void CodeDiffDialog::OnExclude()
+{
+  const auto recorded_symbols = CalculateSymbolsFromProfile();
+  if (m_include.empty() && m_exclude.empty())
+  {
+    m_exclude = recorded_symbols;
+    return;
+  }
+
+  std::vector<Diff> current_diff;
+  if (m_exclude.empty())
+  {
+    m_exclude = recorded_symbols;
+  }
+  else
+  {
+    for (auto& iter : recorded_symbols)
+    {
+      auto pos = lower_bound(m_exclude.begin(), m_exclude.end(), iter.symbol, AddrOP);
+
+      if (pos->symbol != iter.symbol)
+      {
+        current_diff.push_back(iter);
+
+        m_exclude.insert(pos, iter);
+      }
+    }
+    // If there is no include list, we're done.
+    if (m_include.empty())

```

Comment:
I think this check should also apply when in the other block of the `if..else` (i.e. when `if (m_exclude.empty())`). If so, this check can be moved down:

```c++
if (!current_diff.empty() && !m_include.empty())
  RemoveMatchingSymbolsFromIncludes(current_diff);
```

Reply:
Originally I agreed but this check exists to abort in only the parent 'else' case.
Adding a m_include constraint prevents a OnExclude() starting press rather that Including first.